### PR TITLE
fix(streaming): cap the receiver line buffer to the protocol limit

### PR DESCRIPTION
**Pull request by Pyolar**
##### Summary

The streaming receiver keeps input in a line buffer until a newline shows up, and nothing ever bounded how large that buffer can get. The buffer is grown with `buffer_need_bytes()` on every read that does not contain a newline, so a child that simply streams bytes without `\n` makes the parent allocate memory 1:1 with the input. In testing, ~96MB of such input added ~98MB to the parent's RSS. A child that talks the protocol normally sends lines well below the limit, so a connection running away like this means the sender is broken or hostile.

A single protocol line can never exceed `PLUGINSD_LINE_MAX` bytes. This caps the receiver line buffer there and drops the connection (with a log message) when a sender keeps the buffer past the limit.

Adds `stream_receiver_line_buffer_overflow()` in `stream-receiver-internals.h` and a standalone test (built as `stream-receiver-line-cap-test`) that drives the real line primitive directly and covers the accumulation behavior, the exact cap boundary and complete lines.

##### Test Plan

- `stream-receiver-line-cap-test`: OK (no-newline accumulation hits the cap, boundary at `PLUGINSD_LINE_MAX` / `+1`, complete lines never trip it)
- `netdata -c netdata.conf -W unittest`: all tests pass
- Manual: parent on the loopback, child sends newline-free data → connection dropped once the buffer passes the limit, RSS stays flat; normal line-based streaming keeps working

##### Additional Information

Noticed while going through the streaming receiver. The other deferred buffers in the same codebase (e.g. `PLUGINSD_MAX_DEFERRED_SIZE`) already have caps; this path was the odd one out.

<details> <summary>For users: How does this change affect me?</summary>
Parents drop streaming connections that send data without any newline once the accumulated bytes pass the protocol line limit, instead of letting memory grow until the process dies. Normal children are not affected.
</details>

---
## Summary by cubic
Caps the streaming receiver's line buffer at `PLUGINSD_LINE_MAX` bytes so a sender that streams data without newlines no longer makes the parent's memory grow 1:1 with input until OOM. The connection is dropped with a log message once the buffer passes the limit. The cap is checked before the parser consumes a line, so an overlong line that arrives with a newline is caught too, in both compressed and uncompressed streams; normal line-based streaming is unaffected.

**Bug Fixes**
- Adds `stream_receiver_line_buffer_overflow()` and a shared `stream_receiver_overlong_line()` disconnect path for both compressed and uncompressed streams.
- Added `stream-receiver-line-cap-test` covering buffer accumulation, the exact cap boundary, overlong complete lines, and normal lines.

<sup>Written for commit 99c1190e0237202cff66c0772ad9efd0720fb6d3. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/netdata/netdata/pull/23683?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented streaming connections from consuming unbounded memory when incoming protocol lines exceed the allowed size.
  * Oversized lines are now detected during both compressed and uncompressed stream processing.
  * Affected connections are logged and disconnected safely when protocol lines exceed the limit.
  * Lines exactly at the permitted size continue to be accepted.

* **Tests**
  * Added coverage for line-length limits, boundary conditions, partial data, oversized terminated lines, and repeated valid messages.

### CODE REVIEW

File: src/streaming/stream-receiver.c [hunk 1]
```diff
@@ -857,6 +857,18 @@ stream_receive_and_process(struct stream_thread *sth, struct receiver_state *rpt
             *removed = true;
             return -1;
         }
+
+        if(unlikely(stream_receiver_line_buffer_overflow(rpt->thread.line_buffer))) {
```
[the remark this answers is not in the document]
Comment (thiagoftsm) [reply src/streaming/stream-receiver.c:861]: Hello @Pyolar,

Thank you for your contribution.

I have approved the CI to run and verify that everything is okay with your changes.

Please take a look at this specific review.

**Comment by Pyolar** [comment]
Good catch, the check order was wrong. Fixed in 99c1190e02:

- The cap is now checked right after `buffered_reader_next_line()` appends to the line buffer, before `parser_action()` and before the buffer is reset, so an overlong complete line can no longer slip through the parser.
- In the compressed path the same check runs per complete line, plus once after the current decompressed buffer is drained, before `receiver_get_decompressed()` is called again - so a partial line cannot keep growing across decompressed buffers.
- The uncompressed path keeps the check after the loop as well, for the partial line built up across reads.
- Logging and `stream_receiver_remove(..., STREAM_HANDSHAKE_DISCONNECT_BUFFER_OVERFLOW)` moved into `stream_receiver_overlong_line()` so both paths share one copy.

`stream-receiver-line-cap-test` now drives the receiver's check order and adds the two regression cases: an overlong `\n`-terminated complete line (must overflow before the parser reset, and a check ordered after the reset must not see it), and multi-chunk accumulation without a newline (must trip on the chunk that crosses the cap). Short lines, the cap boundary and the existing cases still pass. `netdata -W unittest` passes as well.

**Comment by Pyolar** [comment]
Reduced the new-code duplication that the SonarQube quality gate flagged, with the fix semantics unchanged (5df690ac96):

- The per-line loop is now shared by both read paths in `stream_receiver_process_lines()`. It still checks the cap right after `buffered_reader_next_line()` appends to the line buffer and before `parser_action()`, so the ordering fix lives in one place now.
- The append-then-check pair moved into `stream_receiver_next_line_checked()` so the compressed and uncompressed paths cannot drift apart again.
- `stream-receiver-line-cap-test` keeps the four meaningful cases: a partial line crossing the cap, the exact `PLUGINSD_LINE_MAX` boundary, an overlong `\n`-terminated complete line, and a normal short line. The two near-identical accumulation scenarios and the old-ordering demonstration are gone.

`PLUGINSD_LINE_MAX` is still accepted, anything larger still disconnects with `STREAM_HANDSHAKE_DISCONNECT_BUFFER_OVERFLOW`, and a partial line is still checked between decompressed buffers and across reads. Verified with `stream-receiver-line-cap-test`, `netdata -W unittest`, and end-to-end: overlong complete line and no-newline accumulation both disconnect, normal short lines are still stored and queryable.
